questionpy_server/settings.py

```diff
@@ -96,26 +122,29 @@ def transform_to_path(cls, value: Optional[DirectoryPath]) -> Optional[Directory
 
 
 class Settings(BaseSettings):
+    general: GeneralSettings
     webservice: WebserviceSettings
     worker: WorkerSettings
     cache_package: PackageCacheSettings
     cache_question_state: QuestionStateCacheSettings
     collector: CollectorSettings
 
-    config_files: Tuple[Path, ...] = ()
+    config_files: tuple[Path, ...] = ()
 
     class Config:
         env_prefix = 'qpy_'
+        env_nested_delimiter = '__'
 
         @classmethod
         def customise_sources(
                 cls,
                 init_settings: InitSettingsSource,
-                env_settings: SettingsSourceCallable,
+                env_settings: EnvSettingsSource,
                 # pylint: disable=unused-argument
                 file_secret_settings: SettingsSourceCallable
-        ) -> Tuple[Callable, ...]:
+        ) -> tuple[Callable, ...]:
             if "config_files" in init_settings.init_kwargs:
-                return init_settings, IniFileSettingsSource(init_settings.init_kwargs["config_files"]), env_settings
+                ini_settings = IniFileSettingsSource(init_settings.init_kwargs["config_files"])
+                return init_settings, CombinedSettingsSource(env_settings, ini_settings)
```

Comment:
> customise_sources takes three callables as arguments and returns any number of callables as a tuple. In turn these callables are called to build the inputs to the fields of the settings class.

Warum brauchen wir hier also `CombinedSettingsSource`? Das klingt so, als würde pydantic die zurückgegebenen Sources eh kombinieren.

Reply:
Oh..., ich hatte es so verstanden, dass beim Erstellen der Settings alle Sources durchlaufen werden, bis das Bauen der Settings erfolgreich ist.
Tatsächlich überschreiben die `env_settings` aber die `ini_settings`, sobald man die Reihenfolge tauscht.

Wir können aber mit Hilfe der Klasse das Überschreiben einer `config.ini`-Einstellung mit einer Umgebungsvariable loggen, sodass der Admin sofort sieht, woher die einzelnen Einstellungen stammen. Ein Admin könnte z.B. die erstellte Umgebungsvariable vergessen und nicht verstehen, weshalb sich die Einstellung trotz Änderung in der `config.ini` nicht ändert.

Reply:
Das passiert hier ja noch nicht. Halte ich auch nicht wirklich für nötig. Die Umgebungsvariablen werden ja später in einer Unit-Datei (Systemd-Service) oder docker-compose.yml oder so angegeben. Da ist es gar nicht mal so leicht, eine Umgebungsvariable zu vergessen ^^